### ISSUE TITLE
Interest in a PR that adds TCP keep-alive?

### ISSUE THREAD

**Issue by saethlin** [algesten/ureq#547]
I'm currently working through what to do about a codebase that uses `ureq` and is suffering from what looks like a very occasionally unresponsive server. I'd like to toggle on TCP keepalive to see if that fixes the problem, but currently there is no API for this. `reqwest` has this: https://docs.rs/reqwest/0.11.12/reqwest/blocking/struct.ClientBuilder.html#method.tcp_keepalive and the implementation which is in the middle of being migrated to `hyper-util` looks pretty simple: https://github.com/hyperium/hyper-util/blob/85aade41a080fc4b40a9a53c818680feed14f138/src/client/connect/http.rs#L593-L598

Are you interested in accepting a PR which adds this feature?

**Comment by algesten** [comment]
I looked into this recently in another project, and found that [Rust std `TcpStream` does not have any way of setting keepalive](https://github.com/rust-lang/rust/issues/69774). reqwest (which is based on hyper), pulls in [socket2](https://docs.rs/hyper/0.14.20/src/hyper/server/tcp.rs.html#112-113) to do this, but I'm not keen on adding another dep to ureq _just_ to support keepalive.

I haven't studied hyper in detail, if they have more reasons that keepalive to use `socket2`, but there's also a philosophical difference. Ureq actively tries to keep the dependency tree small, while tokio/hyper has less such constraints.

**Comment by algesten** [comment]
Or in other words: This is the first time anyone raised the issue about `TCP_KEEPALIVE`. If there is strong user demand for it, I could be swayed of increasing the number of deps. I suggest we leave this issue open here to monitor demand and let people weigh in. Hopefully we can close this using just Rust `std` in the future.

**Comment by jsha** [comment]
By the way, my experience of this category of problem is that it is often caused by a firewall that silently drops idle TCP connections (that is, without sending a RST). For clients that keep a connection pool, the connection appears to be open and the next request is sent on it. From the firewall's perspective, it receives a TCP segment for a connection that does not exist, and drops it. The client winds up waiting for a response that will never come. Depending on the OS settings it might wait 30 minutes to an hour before it gets an error.

TCP keepalive is one way to work around this problem. Some others to try for the short term (of course, these have their drawbacks):

 - Set a smallish response timeout, and retry timed-out requests (assuming they are idempotent).
 - Turn off connection pooling.
 - Limit the connection pool to 1 connection per host, and make sure to send at least 1 request per N minutes (where N minutes is less than the observed time after which a connection drops). The request can be an idempotent one, e.g. a GET or HEAD.
 - Implement a [TlsConnector](https://docs.rs/ureq/latest/ureq/trait.TlsConnector.html) that makes a connection, [sets keepalive using socket2](https://docs.rs/socket2/latest/socket2/struct.Socket.html), turns that socket into a `TcpStream`, wraps it with the TLS library of your choice, and hands off the resulting connection to ureq.

**Comment by HookedBehemoth** [comment]
I'd be interested to see this too. I'm unsure if this helps with my use case though.

**Comment by sosthene-nitrokey** [comment]
Interested in this too, see https://github.com/Nitrokey/nethsm-pkcs11/pull/164

**Comment by algesten** [comment]
We should be on the lookout for lightweight crates that solve this problem. Add it under a feature flag.

**Comment by sosthene-nitrokey** [comment]
With the new unstable transport API you can now implement keepalives yourself: https://github.com/Nitrokey/nethsm-pkcs11/blob/c863f31f5cb64238b38ef7086898603ed97c25e2/pkcs11/src/ureq/tcp_connector.rs#L52

**Comment by algesten** [comment]
Yes, we could potentially feature flag and use socket2 to get this already now: https://docs.rs/socket2/0.5.8/socket2/struct.Socket.html#method.set_keepalive
